File: webui/src/lib/roles.ts
/**
 * Amendia `role.*` vocabulary (mirrors the platform contracts / identity seed).
 * Roles are plain strings from `GET /me` — never parsed from the token. These are
 * the ids the UI keys progressive disclosure off (the backend still enforces).
 */
export const ROLE = {
  analyst: "role.payments.ops_analyst",
  approver: "role.payments.ops_approver",
  processOwner: "role.process.owner",
  platformAdmin: "role.platform.admin",
} as const;

export type RoleId = string;

/**
 * Roles that make someone an *operator* (they work the payment-exception flow). The
 * operator nav entries (dashboard / inbox / instances / exceptions) are shown to any
 * of these — so a platform-admin-*only* user (e.g. alex) sees just Administration,
 * while every existing operator persona's nav is unchanged (analyst / approver /
 * process-owner all remain operators). Reads themselves stay role-free server-side;
 * this is purely progressive-disclosure of the nav.
 */
export const OPERATOR_ROLES: string[] = [
  ROLE.analyst,
  ROLE.approver,
  ROLE.processOwner,
];

const ROLE_LABEL: Record<string, string> = {
  [ROLE.analyst]: "Analyst",
  [ROLE.approver]: "Approver",
  [ROLE.processOwner]: "Process owner",
  [ROLE.platformAdmin]: "Platform admin",
};

/** Shape guard for a `role.*` id (mirrors backend `ROLE_ID_RE`). */
const ROLE_ID_RE = /^role\.[a-z0-9_.]+$/;
export function isValidRoleId(role: string): boolean {
  return ROLE_ID_RE.test(role);
}

/** Fallback label for a role with no curated/authored name: last dotted segment, Title Cased. */
export function humanizeRole(role: string): string {
  const tail = role.split(".").pop() ?? role;
  const label = tail.replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());
  return label || role;
}

export function roleLabel(role: string | null | undefined): string {
  if (!role) return "—";
  return ROLE_LABEL[role] ?? humanizeRole(role);
}

/** Short, comma-joined labels for a role set (top-bar chip). */
export function rolesSummary(roles: string[]): string {
  if (!roles.length) return "No roles";
  return roles.map(roleLabel).join(" · ");
}

/** Plain-language, admin-facing descriptions (shown on role rows + assign cards). */
export const ROLE_DESCRIPTION: Record<string, string> = {
  [ROLE.analyst]: "Reviews assessments and drafts repairs and returns at analyst gates.",
  [ROLE.approver]: "Approves results and authorizes payment actions at approver gates.",
  [ROLE.processOwner]: "Authors, validates, and activates process packs in the Registry.",
  [ROLE.platformAdmin]:
    "Full administrative control: manages users, roles, and staged access across the platform.",
};

export function roleDescription(role: string): string {
  return ROLE_DESCRIPTION[role] ?? "";
}

/**
 * The two genuinely code-fixed platform roles (enforced by static route guards, not by any
 * pack). They are always assignable regardless of what packs reference, so the admin picker
 * merges them in even when the roles-in-use query is empty or still loading.
 */
export const PLATFORM_ROLES: string[] = [ROLE.processOwner, ROLE.platformAdmin];

/** One entry the admin role picker renders (a merge of platform roles + pack roles in use). */
export interface AssignableRole {
  id: string;
  label: string;
  description: string;
  isAdmin: boolean;
  /** pack_key@version references, for pack-local roles (absent/empty for platform roles). */
  sources?: string[];
}

/** Minimal shape of the `GET /roles` rows (see `RoleInUse` in the registry service). */
export interface RoleInUseLike {
  role_id: string;
  label?: string | null;
  description?: string | null;
  sources?: string[];
}

/**
 * Build the assignable-role catalog for the admin dialogs: the code-fixed {@link PLATFORM_ROLES}
 * unioned with the roles active packs actually reference (from `GET /roles`), deduped by id.
 * Labels/descriptions prefer the endpoint's authored metadata, then the curated platform maps,
 * then a humanized fallback — so a brand-new pack role is assignable with a sensible name even
 * before anyone authors metadata for it.
 */
export function buildAssignableRoles(rolesInUse: RoleInUseLike[]): AssignableRole[] {
  const out: AssignableRole[] = [];
  const seen = new Set<string>();

  const push = (id: string, authored?: RoleInUseLike) => {
    if (seen.has(id)) return;
    seen.add(id);
    out.push({
      id,
      label: authored?.label || ROLE_LABEL[id] || humanizeRole(id),
      description: authored?.description || ROLE_DESCRIPTION[id] || "",
      isAdmin: isAdminRole(id),
      sources: authored?.sources && authored.sources.length ? authored.sources : undefined,
    });
  };

  // Platform roles first (curated, disclosure order), then pack roles in use.
  for (const id of PLATFORM_ROLES) push(id);
  for (const r of rolesInUse) push(r.role_id, r);
  return out;
}

/** True for the elevated platform-admin role — drives caution treatment. */
export function isAdminRole(role: string): boolean {
  return role === ROLE.platformAdmin;
}

/**
 * Badge variant per role. `platform admin` is rendered with the amber "attention"
 * treatment so it reads as an elevated, handle-with-care grant; the operating roles
 * are neutral outlines.
 */
export function roleBadgeVariant(role: string): "attention" | "outline" {
  return isAdminRole(role) ? "attention" : "outline";
}

File: backend/services/process-registry/app/services/roles.py
# app/services/roles.py
"""Roles in use across active packs.

Role **ids** are always *derived* from active packs' bindings (`hitl.role` +
`executor.role` for human executors) — this is universal and works for seed/API-onboarded
packs that carry no authored metadata. The per-pack ``pack_roles`` sidecar (authored during
onboarding) only *enriches* those ids with a label/description. The two code-fixed platform
roles (``role.process.owner`` / ``role.platform.admin``) are NOT surfaced here — they are a
curated frontend constant merged in at the UI layer.
"""
from __future__ import annotations

from typing import Dict, List

from app.dal.pack_repo import ProcessPackRepository
from app.models.registry import RoleInUse


def _roles_from_bindings(pack: dict) -> List[str]:
    roles: List[str] = []
    for b in pack.get("bindings", []) or []:
        hitl = b.get("hitl") or {}
        if hitl.get("role"):
            roles.append(hitl["role"])
        executor = b.get("executor") or {}
        if executor.get("type") == "human" and executor.get("role"):
            roles.append(executor["role"])
    return roles


async def list_roles_in_use(pack_repo: ProcessPackRepository) -> List[RoleInUse]:
    """Union of role ids referenced by active packs, enriched with sidecar metadata.

    Deduped by role_id; ``sources`` lists the ``pack_key@version`` packs that reference it.
    Metadata (label/description) comes from the first pack whose sidecar carries it."""
    by_id: Dict[str, RoleInUse] = {}

    for pack in await pack_repo.list_active_raw():
        pk, ver = pack.get("pack_key", ""), pack.get("version", "")
        source = f"{pk}@{ver}"
        role_ids = set(_roles_from_bindings(pack))
        if not role_ids:
            continue
        meta = {r.get("role_id"): r for r in await pack_repo.get_pack_roles(pk, ver)}
        for rid in role_ids:
            entry = by_id.get(rid)
            if entry is None:
                entry = RoleInUse(role_id=rid)
                by_id[rid] = entry
            if source not in entry.sources:
                entry.sources.append(source)
            m = meta.get(rid)
            if m and entry.label is None and m.get("label"):
                entry.label = m["label"]
                entry.description = m.get("description") or None

    for entry in by_id.values():
        entry.sources.sort()
    return sorted(by_id.values(), key=lambda r: r.role_id)
